### PR TITLE
able to scale in/out with delete slots feature

### PR DESCRIPTION
**Pull request by cofyc**
### What problem does this PR solve? 

xref: #1348 

### What is changed and how does it work?

- use advanced-statefulset v0.2.1
- able to scale in/out with delete slots feature

### Check List 

Tests 

 - Unit test
 - E2E test

### Does this PR introduce a user-facing change?:
 
 ```release-note
able to scale in/out with delete slots feature of advanced statefulset
 ```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -142,3 +142,5 @@ replace k8s.io/legacy-cloud-providers => k8s.io/legacy-cloud-providers v0.0.0-20
 replace k8s.io/node-api => k8s.io/node-api v0.0.0-20190918163711-2299658ad911
 
 replace github.com/uber-go/atomic => go.uber.org/atomic v1.5.0
+
+replace github.com/pingcap/advanced-statefulset => /data/go/src/github.com/pingcap/advanced-statefulset
```
Comment (gregwebs) [inline go.mod]: probably don't want to commit this

Comment (cofyc) [reply go.mod]: this is for local development only, I'll remove this

File: tests/actions.go [hunk 2]
```diff
@@ -393,15 +397,19 @@ func (oi *OperatorConfig) OperatorHelmSetString(m map[string]string) string {
 		"scheduler.kubeSchedulerImageName":           oi.SchedulerImage,
 		"controllerManager.logLevel":                 oi.LogLevel,
 		"scheduler.logLevel":                         "4",
-		"controllerManager.replicas":                 "2",
-		"scheduler.replicas":                         "2",
```
Comment (cofyc) [inline tests/actions.go]: make it configurable, in some tests, we can use 1 replica to consumes fewer resources

File: tests/actions.go [hunk 3]
```diff
@@ -1326,15 +1340,23 @@ func (oa *operatorActions) CheckUpgradeOrDie(ctx context.Context, info *TidbClus
 func (oa *operatorActions) DeployMonitor(info *TidbClusterConfig) error { return nil }
 func (oa *operatorActions) CleanMonitor(info *TidbClusterConfig) error  { return nil }
 
-func getMemberContainer(kubeCli kubernetes.Interface, namespace string, memberName string) (*corev1.Container, bool) {
-	name := fmt.Sprintf("%s-%d", memberName, 0)
-	pod, err := kubeCli.CoreV1().Pods(namespace).Get(name, metav1.GetOptions{})
+// getMemberContainer gets member container
+func getMemberContainer(kubeCli kubernetes.Interface, namespace, tcName, component string) (*corev1.Container, bool) {
```
Comment (cofyc) [inline tests/actions.go]: instead of fetching the `-0` pod, we list the pods and find the first one.

File: tests/dt.go [hunk 4]
```diff
@@ -63,7 +63,7 @@ func (oa *operatorActions) LabelNodes() error {
 	}
 
 	for i, node := range nodes.Items {
-		err := wait.Poll(3*time.Second, time.Minute, func() (bool, error) {
+		err := wait.PollImmediate(3*time.Second, time.Minute, func() (bool, error) {
```
Comment (cofyc) [inline tests/dt.go]: no need to wait at the first try, this reduces time

File: pkg/controller/tidb_control.go [hunk 5]
```diff
@@ -41,7 +41,7 @@ type DBInfo struct {
 // TiDBControlInterface is the interface that knows how to manage tidb peers
 type TiDBControlInterface interface {
 	// GetHealth returns tidb's health info
-	GetHealth(tc *v1alpha1.TidbCluster) map[string]bool
+	GetHealth(tc *v1alpha1.TidbCluster, ordinal int32) (bool, error)
```
Comment (cofyc) [inline pkg/controller/tidb_control.go]: tidb member controller will decide which member we need to request the health status

File: pkg/manager/member/scaler.go [hunk 6]
```diff
@@ -104,3 +113,91 @@ func ordinalPVCName(memberType v1alpha1.MemberType, setName string, ordinal int3
 func ordinalPodName(memberType v1alpha1.MemberType, tcName string, ordinal int32) string {
 	return fmt.Sprintf("%s-%s-%d", tcName, memberType, ordinal)
 }
+
+func validateScaling(actual *apps.StatefulSet, desired *apps.StatefulSet) error {
+	if !features.DefaultFeatureGate.Enabled(features.AdvancedStatefulSet) {
+		return nil
+	}
+	// Following cases are not supported yet.
+	// TODO support it or validate in validation phase
```
Comment (cofyc) [inline pkg/manager/member/scaler.go]: some scaling operations are not allowed because it has to delete/create pods. I decided to implement in the future. There are two ways to solve this issue:

1) avoid in validation phase
2) support it by scaling out first and then scaling in, e.g.
   -`replicas 5, delete slots [0]` is changed to `replicas 4, delete slots[1]`, we need to delete pod 1 and 5 and create pod 4.

File: pkg/features/features.go [hunk 7]
```diff
@@ -45,9 +47,15 @@ type FeatureGate interface {
 	AddFlag(flagset *flag.FlagSet)
 	// Enabled returns true if the key is enabled.
 	Enabled(key string) bool
+	// Set parses and stores flag gates for known features
+	// from a string like feature1=true,feature2=false,...
+	Set(value string) error
+	// SetFromMap stores flag gates for enabled features from a map[string]bool
+	SetFromMap(m map[string]bool)
```
Comment (cofyc) [inline pkg/features/features.go]: able to change features settings in unit tests

File: pkg/manager/member/pd_scaler.go [hunk 8]
```diff
@@ -39,27 +39,42 @@ func NewPDScaler(pdControl pdapi.PDControlInterface,
...
+	}
+	return nil
+}
+
 func (psd *pdScaler) ScaleOut(tc *v1alpha1.TidbCluster, oldSet *apps.StatefulSet, newSet *apps.StatefulSet) error {
+	ordinal, replicas, deleteSlots, err := scaleOne(oldSet, newSet)
+	if err != nil {
+		return err
+	}
+	resetReplicas(newSet, oldSet)
```
Comment (cofyc) [inline pkg/manager/member/pd_scaler.go]: I've updated the scaling logic like this:

- get desired replicas and delete slot
- reset the new statefulset
- if our scaling logic succeeds, set desired replicas and delete slots, otherwise, return an error

File: pkg/manager/member/scaler.go [hunk 9]
```diff
@@ -104,3 +113,91 @@ func ordinalPVCName(memberType v1alpha1.MemberType, setName string, ordinal int3
...
+// return the last ordinal in the list, -1 if the list is empty
+func lastOrdinalInList(list []int) int {
+	if len(list) == 0 {
+		return -1
+	}
+	return list[len(list)-1]
+}
+
+// scaleOne scales actual set to desired set by deleting or creating a replica
+func scaleOne(actual *apps.StatefulSet, desired *apps.StatefulSet) (int32, int32, sets.Int, error) {
```
Comment (cofyc) [inline pkg/manager/member/scaler.go]: this function calculates desired `replicas` and `delete slots` from actual/desired stateful sets by allowing one pod to be deleted or created

File: charts/tidb-operator/templates/advanced-statefulset-rbac.yaml [hunk 10]
```diff
@@ -36,6 +36,7 @@ rules:
 - apiGroups:
   - ''
   resources:
+  - 'events'
```
Comment (cofyc) [inline charts/tidb-operator/templates/advanced-statefulset-rbac.yaml]: the controller needs this permission to report events

File: pkg/util/util.go [hunk 11]
```diff
@@ -87,20 +93,61 @@ func GetOrdinalFromPodName(podName string) (int32, error) {
 	return int32(ordinalInt), nil
 }
 
-func GetNextOrdinalPodName(podName string, ordinal int32) string {
```
Comment (cofyc) [inline pkg/util/util.go]: unused function

File: pkg/manager/member/scaler.go [hunk 12]
```diff
@@ -104,3 +113,91 @@ func ordinalPVCName(memberType v1alpha1.MemberType, setName string, ordinal int3
...
+	if *actual.Spec.Replicas == *desired.Spec.Replicas {
+		// no scaling
+		if actualDeleteSlots.Equal(desiredDeleteSlots) {
+			return nil
+		}
+	} else if *actual.Spec.Replicas > *desired.Spec.Replicas {
+		// scaling in
+		if actualDeleteSlots.Equal(desiredDeleteSlots) || desiredDeleteSlots.IsSuperset(actualDeleteSlots) {
+			return nil
+		}
```
[the remark this answers is not in the document]
Comment (cofyc) [reply pkg/manager/member/scaler.go]: you're right, your suggestion method is simpler

File: pkg/manager/member/scaler.go [hunk 13]
```diff
@@ -104,3 +113,91 @@ func ordinalPVCName(memberType v1alpha1.MemberType, setName string, ordinal int3
...
+			actualDeleteSlots.Delete(slot)
+			return int32(slot), *actual.Spec.Replicas + 1, actualDeleteSlots, nil
+		} else if !desiredDeleteSlots.Equal(desiredDeleteSlots) {
+			// unreachable because this is prohibited by validateScaling
+			return -1, -1, nil, fmt.Errorf("unsupported scaling operation for set %s/%s, actual (replicas: %d, delete slots: %v), desired (replicas: %d, delete slots: %v)", actual.Namespace, actual.Name, *actual.Spec.Replicas, actualDeleteSlots, *desired.Spec.Replicas, desiredDeleteSlots)
+		}
+		// actualDeleteSlots equals to desiredDeleteSlots
+		ordinalList := helper.GetDesiredPodOrdinals(int(*actual.Spec.Replicas), actual).List()
+		return int32(lastOrdinalInList(ordinalList)) + 1, *actual.Spec.Replicas + 1, actualDeleteSlots, nil
+	}
```
[the remark this answers is not in the document]
Comment (cofyc) [reply pkg/manager/member/scaler.go]: yes, after this PR (https://github.com/pingcap/tidb-operator/pull/1389), when webhook is enabled `scaleOne` can be skipped. for now, I need to update convert current logic to work with advanced statefulset

[the remark this answers is not in the document]
Comment (aylei) [reply pkg/manager/member/scaler.go]: IMHO, the logic could be more clear if we trigger actions based on the comparison of `DesiredPodOrdinals` of current and next state, that way, we can easily tell that the reconling logic is irrelevant to previous states and could behave consistently across multiple round of reconcilings.

[the remark this answers is not in the document]
Comment (cofyc) [reply pkg/manager/member/scaler.go]: it makes sense, then some unsupported desired state can be easily achieved in this way. the only question is if some pods need to be created and some pods needs, which operation should be performed first? for safety, I'll create the pods first.

the only validation is to check the format of delete slots value to warn the users if they set malformed value

File: pkg/manager/member/utils.go [hunk 14]
```diff
@@ -131,8 +133,32 @@ func encode(obj interface{}) (string, error) {
...
+		}
+	}
+	if actual != nil {
+		for k, v := range actual {
+			actualCopy[k] = v
+		}
+	}
+	delete(desiredCopy, LastAppliedConfigAnnotation)
+	delete(actualCopy, LastAppliedConfigAnnotation)
+	return !reflect.DeepEqual(desiredCopy, actualCopy)
```
[the remark this answers is not in the document]
Comment (cofyc) [reply pkg/manager/member/utils.go]: removed in 1c5832eae80ba8d5633e844aa3672860e27f9b26

File: pkg/manager/member/scaler.go [hunk 15]
```diff
@@ -104,3 +113,91 @@ func ordinalPVCName(memberType v1alpha1.MemberType, setName string, ordinal int3
...
+	} else {
+		// scaling out
+		if actualDeleteSlots.Equal(desiredDeleteSlots) || actualDeleteSlots.IsSuperset(desiredDeleteSlots) {
+			return nil
+		}
+	}
+	return fmt.Errorf("unsupported scaling operation for sts %s/%s (actual: %d, %v; desired: %d, %v)", actual.Namespace, actual.Name,
+		*actual.Spec.Replicas, actualDeleteSlots.List(),
+		*desired.Spec.Replicas, desiredDeleteSlots.List(),
+	)
```
[the remark this answers is not in the document]
Comment (cofyc) [reply pkg/manager/member/scaler.go]: implemented in d72e567c5422b9d57ee7bb2ea02233bf9e520ae8

File: pkg/manager/member/scaler.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (cofyc) [reply pkg/manager/member/scaler.go]: validatingScaling is not needed anymore

File: pkg/manager/member/scaler.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (cofyc) [reply pkg/manager/member/scaler.go]: refactored in 20c1f8cfddcdf94ce4355429d74fe19e1621d018

File: pkg/manager/member/pd_scaler.go [hunk 16]
```diff
@@ -39,27 +39,36 @@ func NewPDScaler(pdControl pdapi.PDControlInterface,
...
+	if *newSet.Spec.Replicas > *oldSet.Spec.Replicas {
+		return psd.ScaleOut(tc, oldSet, newSet)
+	} else if *newSet.Spec.Replicas < *oldSet.Spec.Replicas {
+		return psd.ScaleIn(tc, oldSet, newSet)
+	}
+	return nil
+}
+
 func (psd *pdScaler) ScaleOut(tc *v1alpha1.TidbCluster, oldSet *apps.StatefulSet, newSet *apps.StatefulSet) error {
+	_, ordinal, replicas, deleteSlots := scaleOne(oldSet, newSet)
```
Comment (cofyc) [inline pkg/manager/member/pd_scaler.go]: recalculating again because of less code needs to be changed (otherwise ScaleOut/ScaleIn function signatures need to change and a lot of tests must be updated)

File: pkg/manager/member/tikv_scaler.go [hunk 17]
```diff
@@ -41,42 +41,53 @@ func NewTiKVScaler(pdControl pdapi.PDControlInterface,
...
+	if scaling > 0 {
+		return tsd.ScaleOut(tc, oldSet, newSet)
+	} else if scaling < 0 {
+		return tsd.ScaleIn(tc, oldSet, newSet)
+	}
+	return nil
+}
+
 func (tsd *tikvScaler) ScaleOut(tc *v1alpha1.TidbCluster, oldSet *apps.StatefulSet, newSet *apps.StatefulSet) error {
+	_, ordinal, replicas, deleteSlots := scaleOne(oldSet, newSet)
```
Comment (cofyc) [inline pkg/manager/member/tikv_scaler.go]: recalculating again because of less code needs to be changed (otherwise ScaleOut/ScaleIn function signatures need to change and a lot of tests must be updated)

File: pkg/manager/member/tikv_upgrader.go [hunk 18]
```diff
@@ -87,7 +90,18 @@ func (tku *tikvUpgrader) Upgrade(tc *v1alpha1.TidbCluster, oldSet *apps.Stateful
...
+	if features.DefaultFeatureGate.Enabled(features.AdvancedStatefulSet) {
+		maxReplicaCount, deleteSlots = helper.GetMaxReplicaCountAndDeleteSlots(int(*oldSet.Spec.Replicas), helper.GetDeleteSlots(oldSet))
+	} else {
+		maxReplicaCount = int(tc.TiKVStsActualReplicas())
+	}
+	for i := int32(maxReplicaCount) - 1; i >= 0; i-- {
+		if deleteSlots != nil && deleteSlots.Has(int(i)) {
+			glog.Infof("tikv pod %s/%s is in delete slots, skip", ns, TikvPodName(tcName, i))
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (cofyc) [reply pkg/manager/member/tikv_upgrader.go]: implemented in 24988e81e8f705548bbf972244a95c02f24162ba

File: pkg/manager/member/scaler.go [hunk 19]
```diff
@@ -104,3 +113,45 @@ func ordinalPVCName(memberType v1alpha1.MemberType, setName string, ordinal int3
...
 	return fmt.Sprintf("%s-%s-%d", tcName, memberType, ordinal)
 }
+
+// scaleOne calculates desired replicas and delete slots from actual/desired
+// stateful sets by allowing only one pod to be deleted or created
+// it returns following values:
+// - scaling:
+//   - 0: no scaling required
+//   - 1: scaling out
+//   - -1: scaling in
```
Comment (aylei) [inline pkg/manager/member/scaler.go]: Could we define enums for this?

File: pkg/manager/member/scaler_test.go [hunk 20]
```diff
@@ -158,3 +163,188 @@ func newFakeGeneralScaler() (*generalScaler, cache.Indexer, *controller.FakePVCC
 	return &generalScaler{pvcLister: pvcInformer.Lister(), pvcControl: pvcControl},
 		pvcInformer.Informer().GetIndexer(), pvcControl
 }
+
+func TestScaleOne(t *testing.T) {
```
Comment (aylei) [inline pkg/manager/member/scaler_test.go]: Could we add a case for "scale tidb to 0"?

File: pkg/manager/member/scaler.go [hunk 21]
```diff
@@ -104,3 +113,45 @@ func ordinalPVCName(memberType v1alpha1.MemberType, setName string, ordinal int3
...
+// - ordinal: pod ordinal to create or delete
+// - replicas/deleteSlots: desired replicas and deleteSlots by allowing only one pod to be deleted or created
+func scaleOne(actual *apps.StatefulSet, desired *apps.StatefulSet) (scaling int, ordinal int32, replicas int32, deleteSlots sets.Int32) {
+	actualDesiredPodOrdinals := helper.GetPodOrdinals(*actual.Spec.Replicas, actual)
+	desiredDesiredPodOrdinals := helper.GetPodOrdinals(*desired.Spec.Replicas, desired)
+	additions := desiredDesiredPodOrdinals.Difference(actualDesiredPodOrdinals)
+	deletions := actualDesiredPodOrdinals.Difference(desiredDesiredPodOrdinals)
```
Comment (aylei) [inline pkg/manager/member/scaler.go]: Is the order of the diff result guaranteed? or random?

Comment (cofyc) [reply pkg/manager/member/scaler.go]: `List()` will sort the set members in ascending order

File: pkg/manager/member/scaler_test.go (hunk 20 above)
Comment (cofyc) [reply pkg/manager/member/scaler_test.go]: fixed some corner cases and add more tests in c8eb4058964cd5b7b6088b6d5221c0b17c8929f2

File: pkg/manager/member/scaler.go (hunk 19 above)
Comment (cofyc) [reply pkg/manager/member/scaler.go]: I don't think it's necessary because it's used internally and I actually uses `> 0` and `< 0` to decide whether to scale in or scale out.
